### PR TITLE
Noetic support

### LINKED ISSUE #29: Noetic release
Would you mind releasing these packages also for the newest distribution ROS Noetic? :)

### LINKED ISSUE #34: Some of the dependencies are not compatible with python3
Like "import StringIO" in File "~/smach_viewer/smach_viewer.py", line 44, in <module> import StringIO

I would deeply appreciate that if you can update this lib to the noetic version.

### LINKED ISSUE #36: ImportError: No module named 'StringIO' for python3
https://github.com/ros-visualization/executive_smach_visualization/blob/0d0d4321e1194061c63a2fdc42029a29baa10440/smach_viewer/scripts/smach_viewer.py#L44
This line gives `ModuleNotFoundError: No module named 'StringIO'` in python3

From [https://docs.python.org/3.0/whatsnew/3.0.html](url) 

> The StringIO and cStringIO modules are gone. Instead, import the io module and use  io.StringIO or io.BytesIO for text and data respectively.

possible solution: 
```python
try:
    import StringIO 
except ImportError:
    from io import StringIO 
```

### PR DESCRIPTION
**Pull request by InigoMoreno**
I have a working version of smach_viewer in noetic!

I have refactored a lot of the code (especially on the xdot library), so it might be a bit difficult to read through the changes. 

The rosdeps are not correctly defined, as not all of them are available (I will need to do a PR of ros_distro). To run the repo you need to install the following manually (not sure which are necessary):
```bash
 apt install gir1.2-gtk-3.0 python3-gi python3-gi-cairo python3-numpy graphviz
 python3 -m pip install xdot wxpython
 ```

Also, I can't create a new branch on the repo, so I set the base of the fork to melodic-devel, but Ideally, a new noetic-devel branch should be created and the base of this fork changed.

Fixes #29
Fixes #34
Fixes #36

### CODE REVIEW

**Comment by InigoMoreno** [comment]
I have been looking into the deps and I believe only these two are missing from the rosdeps:
```
apt install python3-gi-cairo
 python3 -m pip install xdot
```

**Comment by Timple** [comment]
> I have been looking into the deps and I believe only these two are missing from the rosdeps:
> 
> ```
> apt install python3-gi-cairo
>  python3 -m pip install xdot
> ```

Did you request them to be added there?

File: smach_viewer/package.xml [hunk 1]
```diff
@@ -25,14 +25,22 @@
...
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-wxgtk4.0</exec_depend>
+  <exec_depend>graphviz</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 2">python-gi</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-gi</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 2">python-rospkg</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-rospkg</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 2">python-qt5-bindings</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-qt5-bindings</exec_depend>
+  <exec_depend>python-gi-cairo</exec_depend>
+  <!-- <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-gi-cairo</exec_depend> -->
```
Comment (Timple) [inline smach_viewer/package.xml]: This is actually the right syntax. Why is it commented?

Comment (InigoMoreno) [reply smach_viewer/package.xml]: Because the actual key did not exist, but I see you created a pull request for that.

Comment (InigoMoreno) [reply smach_viewer/package.xml]: Because the python3-gi-cairo key did not exist in ros-repo, but I see that you added them in a PR. Thanks for that!

Comment (Timple) [reply smach_viewer/package.xml]: No problem, xdot still pending: https://github.com/ros/rosdistro/pull/32004

**Comment by Timple** [comment]
Xdot got merged. Can somebody trigger CI again?

**Comment by Timple** [comment]
Seems like kinetic should be removed from CI

**Comment by oss-vineet** [comment]
> I have a working version of smach_viewer in noetic!
> 
> I have refactored a lot of the code (especially on the xdot library), so it might be a bit difficult to read through the changes. 
> 
> The rosdeps are not correctly defined, as not all of them are available (I will need to do a PR of ros_distro). To run the repo you need to install the following manually (not sure which are necessary):
> ```bash
>  apt install gir1.2-gtk-3.0 python3-gi python3-gi-cairo python3-numpy graphviz
>  python3 -m pip install xdot wxpython
>  ```
> 
> Also, I can't create a new branch on the repo, so I set the base of the fork to melodic-devel, but Ideally, a new noetic-devel branch should be created and the base of this fork changed.
> 
> Fixes #29
> Fixes #34
> Fixes #36 

In addition, [from here](https://stackoverflow.com/questions/71369726/no-module-named-gi) I needed the following:
```bash
sudo apt install libxt-dev libgirepository1.0-dev

pip install pycairo PyGObject
```

**Comment by Timple** [comment]
Since the melodic branch has changes inspired by this PR to fix noetic, can this PR now be closed?

Because it has major conflicts and it suggests that noetic is not supported :slightly_smiling_face:

**Comment by MatthijsBurgh** [comment]
Agreeing with @Timple to close this PR.

**Pull request closed without merging**